src/decompress.cpp

```diff
@@ -58,22 +179,27 @@ auto decompress(std::span<const std::byte> src, std::span<std::byte> dst)
         return DecompressStatus::SrcTooSmall;
       }
 
-      if (dst.size() < len) {
+      if (dst.size() - static_cast<std::size_t>(dst_written) < len) {
         return DecompressStatus::DstTooSmall;
       }
 
-      std::copy_n(src_bits.byte_data(), len, dst.begin());
+      std::copy_n(src_bits.byte_data(), len, dst.begin() + dst_written);
```

Comment:
It will be UB if the pointer goes off the front of the original buffer.

```
buf = array<byte, 8>{};
buf: [-|-|-|-|-|-|-|-]

dst = span{buf}
dst: [-|-|-|-|-|-|-|-]

dst = dst | drop(4)
dst:         [-|-|-|-]

it1 = dst.begin()
buf: [-|-|-|-|-|-|-|-]
it1:          ^ 

it2 = it1 - 2
buf: [-|-|-|-|-|-|-|-]
it2:      ^

```

Maybe it would make sense to create a `destination_buffer` type that handles this logic and catches potential UB.

As for how to define the destination buffer, that could depend on how you want to handle growth.

Would it act like a `std::vector` where it allocates a new buffer and copies elements over? In this case, maybe `destination_buffer` stores a pointer and 2 integers.

Or would it maintain a list of chunks so growth doesn't require copying elements from the old buffer to a new buffer? I'm not sure which approach is better, but it would probably help to abstract this part so the rest of the decompression implementation is independent.